File: AspNetMvcUnleashed/0672329980_source/Chapter04/CS/MvcApplication1/MvcApplication1/MyViewEngines/SimpleView.cs
using System.IO;
using System.Text.RegularExpressions;
using System.Web.Mvc;

namespace MvcApplication1.MyViewEngines
{
    public class SimpleView : IView
    {
        private string _viewPhysicalPath;

        public SimpleView(string viewPhysicalPath)
        {
            _viewPhysicalPath = viewPhysicalPath;
        }
        
        #region IView Members

        public void Render(ViewContext viewContext, TextWriter writer)
        {
            // Load file
            string rawContents = File.ReadAllText(_viewPhysicalPath);

            // Perform replacements
            string parsedContents = Parse(rawContents, viewContext.ViewData);

            // Write results to HttpContext
            writer.Write(parsedContents);
        }

        #endregion

        public string Parse(string contents, ViewDataDictionary viewData)
        {
            return Regex.Replace(contents, "\\{(.+)\\}",m => GetMatch(m, viewData));
        }

        protected virtual string GetMatch(Match m, ViewDataDictionary viewData)
        {
            if (m.Success)
            {
                string key = m.Result("$1");
                if (viewData.ContainsKey(key))
                {
                    return viewData[key].ToString();
                }
            }
            return string.Empty;
        }
    
    }
}

File: AspNetMvcUnleashed/0672329980_source/Chapter04/VB/MvcApplication1/MvcApplication1/MyViewEngines/SimpleView.vb
Imports System.IO

Public Class SimpleView
    Implements IView

    Private _viewPhysicalPath As String

    Public Sub New(ByVal viewPhysicalPath As String)
        _viewPhysicalPath = viewPhysicalPath
    End Sub

#Region "IView Members"

	Public Sub Render(ByVal viewContext As ViewContext, ByVal writer As TextWriter) Implements IView.Render
        ' Load file
        Dim rawContents As String = File.ReadAllText(_viewPhysicalPath)

        ' Perform replacements
        Dim parsedContents As String = Parse(rawContents, viewContext.ViewData)

        ' Write results to HttpContext
        writer.Write(parsedContents)
    End Sub

#End Region

    Public Function Parse(ByVal contents As String, ByVal viewData As ViewDataDictionary) As String
        Return Regex.Replace(contents, "\{(.+)\}", Function(m) GetMatch(m, viewData))
    End Function

    Protected Overridable Function GetMatch(ByVal m As Match, ByVal viewData As ViewDataDictionary) As String
        If m.Success Then
            Dim key As String = m.Result("$1")
            If viewData.ContainsKey(key) Then
                Return viewData(key).ToString()
            End If
        End If
        Return String.Empty
    End Function

End Class
